### ISSUE TITLE
resampling is only implemented for RGB images, not for segmentation masks

### ISSUE THREAD

**Issue by YoniSchirris** [NKI-AI/dlup#141]
**Describe the bug**
- When writing a `.tiff` with the `dlup.writers.TifffileImageWriter.from_tile_iterator` of a 2d segmentation mask with a limited number of classes (e.g. `0,1,2,3,4`), `np.unique(np.array(SlideImage.from_file_path('').get_thumbnail(size)))` may produce different output depending on the given `size`, hallucinating additional classes.
- Even when reading at the exact level that the tiff was written, new classes may appear
- Likely, many class-assigned pixels are wrong due to interpolation

**To Reproduce**
Tried it with my environment and @AjeyPaiK 's environment, on tiffs generated from completely distinct pipelines, and we observe the exact same behavior.

In the example below, the model only gives classes 0,1,2,3,4.
```
Python 3.9.12 (main, Apr  5 2022, 06:56:58)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import dlup
>>> import numpy as np
>>> b = dlup.SlideImage.from_file_path('/home/y.schirris/stils_pipeline_yoni/data/output/n4plus/2023-01-09-1/IRBd22_172_n4plus/98273/CF42 T14-61352 A3 #233 HE/seg.tiff')
>>> np.unique(np.array(b.get_thumbnail(1000)))
array([0, 1, 2, 3, 4, 5], dtype=uint8)
>>> np.unique(np.array(b.get_thumbnail(3000)))
array([0, 1, 2, 3, 4, 5, 6], dtype=uint8)
>>> np.unique(np.array(b.get_thumbnail(4000)))
array([0, 1, 2, 3, 4, 5], dtype=uint8)
>>> np.unique(np.array(b.get_thumbnail(6000)))
array([0, 1, 2, 3, 4, 5, 6], dtype=uint8)
```

**Expected behavior**
The tiff writer and dlup slide handler should have an option to state it's a mask, such that the values are treated as categorical variables. This needs to happen in, likely, many places. I will give some suggestions below.

Likely this is as simple as having the functions check the dimensionality of the array that's written and changing its behavior accordingly.

**Suggested changes**

`dlup.SlideImage`
Any `RGB` hardcoding and any `Lanzcos` resampling hardcoding should be replaced by checking the dimensionality and adjusting accordingly

`dlup.TifffileImageWriter`
has `if not is_rgb: output = output[..., 0]`, and

```
tile = get_tile(page, coordinates[::-1], size[::-1])[0]
vips_tile = numpy_to_vips(tile).resize(1 / scale)
output = vips_to_numpy(vips_tile)
```

but this seems not to be enough, because we noticed that if we read at the base level of the segmentation tiff we still found a hallucinated class (and thus there was some continuous interpolation/resampling).

So we need to see where all resizing happens when writing a pyramidal file and make sure that this resizing is done properly for categorical segmentation masks.

**Environment**
dlup version: `0.3.21-dev0`
How installed: github clone & pip install
Python version: 3.9.12
Operating System: linux

**Additional context**

**Comment by AjeyPaiK** [comment]
[SlideImage](https://github.com/NKI-AI/dlup/blob/e990dd0126286ee785a80fede99c1d1d588b898d/dlup/_image.py#L85) expects the WSI to be an instance of the `AbstractSlideBackend` class. I think the reason why we found a hallucinated class at the base level of the segmentation tiff is that we used the [get_thumbnail()](https://github.com/NKI-AI/dlup/blob/e990dd0126286ee785a80fede99c1d1d588b898d/dlup/_image.py#L273) function from the `SlideImage`. 

This in turn, calls the `get_thumbnail()` method within `AbstractSlideBackend`. [If we look at it](https://github.com/NKI-AI/dlup/blob/e990dd0126286ee785a80fede99c1d1d588b898d/dlup/experimental_backends/common.py#L163), the sampling is hardcoded to `Lanzcos`. 

@jonasteuwen, we should modify this hardcoding as per @YoniSchirris's suggestion and make the resampling `nearest` if the tiff image is a categorical mask.

**Comment by jonasteuwen** [comment]
I was working on a PR to do this, together with the utility to make jsons from masks. Look at that one.Sent from my iPhoneOn Feb 24, 2023, at 12:31 AM, Ajey Pai K ***@***.***> wrote:﻿
SlideImage expects the WSI to be an instance of the AbstractSlideBackend class. I think the reason why we found a hallucinated class at the base level of the segmentation tiff is that we used the get_thumbnail() function from the SlideImage.
This in turn, calls the get_thumbnail() method within AbstractSlideBackend. If we look at it, the sampling is hardcoded to Lanzcos.
@jonasteuwen, we should modify this hardcoding as per @YoniSchirris's suggestion and make the resampling nearest if the tiff image is a categorical mask.

**Comment by AjeyPaiK** [comment]
I confirmed that it is indeed a tiff writer problem and not a get_thumbnail problem. I loaded my tiff mask from earlier using rasterio and that also shows the same problem. I should only have 0,1,2,3 in my mask.

![image](https://user-images.githubusercontent.com/26798611/220996505-5fb7f415-0a11-4b94-a37f-9889a4f92059.png)

**Comment by jonasteuwen** [comment]
I think this is now addressed in #138. Can you verify? You can set `interpolator` when creating the `SlideImage`.

**Comment by YoniSchirris** [comment]
@jonasteuwen I believe the `TifffileImageWriter`, when writing a pyramidal file, should also do some interpolation/subsampling/resampling since it writes x2 dowsampled levels. 

I can't find where it does this, though. 

I assume that
- the highest resolution page is written with the raw segmentation output, which should not have any resampling issues
- the pyramidal writing requires some interpolation, so in
```
for level in range(0, n_subresolutions):
                tiff_reader = tifffile.TiffReader(temp_filename)
                page = tiff_reader.pages[level]
                tile_iterator = _tile_iterator_from_page(
                    page, self._tile_size, shapes[level], scale=2, is_rgb=is_rgb
                )
                self._write_page(
                    tiff_writer,
                    tile_iterator=tile_iterator,
                    level=level + 1,
                    compression=_compression,
                    shapes=shapes,
                    is_rgb=is_rgb,
                    subfiletype=1,
                    software=software,
                )
                tiff_reader.close()
```
- Where i expect the `_tile_iterator_from_page` to somehow set how tiles are read from the page and resized
- In which the `grid` just sets the coordinates, and i assume the tile is actually retrieved with `get_tile`
- But i can't find where any resizing is happening

**Comment by jonasteuwen** [comment]
Implemented in #144.
